Encapsulation rework for simplified multi-node operations

tgraf opened the pull request:
This  is a major rework of the encapsulation mode and drastically simplifies operations:
 - No longer requires any network configuration, the IP of the VM/host is automatically used as tunnel endpoint across the mesh
 - There is no longer a need to configure any routes for the container prefixes in the cloud network or the underlying fabric
 - The node prefix to node ip mapping is automatically derived from the Kubernetes PodCIDR

aanm commented on pkg/bpf/map.go:
```diff
@@ -106,6 +106,10 @@ type Map struct {
 	lock sync.RWMutex
+
+	// NonPersistent is true if the map does not contain persistent data
+	// and should be removed on startup.
```
Can you write the name of 1 or 2 maps that don't contain persistent data?

aanm replied:
not addressed?

tgraf replied:
`pkg/bpf/map` is generic and not cilium specific at all. It is already being used by projects outside of Cilium. Using Cilium specific examples would be confusing. Is the description not clear?

aanm replied:
Yes, I thought this was cilium specific and that you were removing maps on startup. If it's generic I'm good.

aanm commented on pkg/maps/tunnel/tunnel.go:
```diff
@@ -0,0 +1,121 @@
+
+// Must be in sync with struct endpoint_key in <bpf/lib/common.h>
+type tunnelEndpoint struct {
+	IP     types.IPv6
```
`// represents both IPv6 and IPv4`?

aanm commented on pkg/maps/tunnel/tunnel.go:
```diff
@@ -0,0 +1,121 @@
+	return mapInstance.Dump(dumpParser, dumpCallback)
+}
+
+func init() {
```
Since this function is likely called first, can we put this function on top?

aanm commented on contrib/vagrant/start.sh:
```diff
@@ -264,7 +264,7 @@ function write_cilium_cfg() {
 
     if [[ "${IPV4}" -eq "1" ]]; then
-        cilium_options+=" --ipv4-range 10.${master_ipv4_suffix}.0.1"
+        cilium_options+=" --ipv4-range 10.${master_ipv4_suffix}.0.0/16"
```
I'm assuming `10.x.0.1/16` will work as fine, right?

aanm replied:
addressed?

aanm commented on daemon/daemon.go:
```diff
@@ -382,20 +383,19 @@ func (d *Daemon) useK8sNodeCIDR(nodeName string) error {
-		return err
+
+	if ip.To4() != nil {
+		nodeaddress.SetIPv4AllocRange(ipnet)
```
Do you want me to replace `nodeaddress.SetNodeAddress(ipv6NodeAddress, ciliumIPv4.NodeIP().String(), "")` in https://github.com/cilium/cilium/pull/1039/commits/159237e067588cd425c2169e74bc95550d33ccaf with this?

tgraf replied:
Yes, 159237e should be rebased on top of this.

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -15,180 +15,189 @@
-	IPv4Route = *v4.IPNet(32)
+	for _, a := range addr {
+		if a.Scope == unix.RT_SCOPE_UNIVERSE {
+			if len(a.IP) >= 4 {
```
Out of curiosity, when is this less than 4?

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -15,180 +15,189 @@
-	}
+// GetIPv4 returns the IPv4 address of the node
+func GetIPv4() net.IP {
+	return ipv4Address
```
Why not get rid of the Getters and Setters and access the IPs/IPNets directly?

tgraf replied:
I had direct access at first but then figured that having Get and Set function allows to do validation as is now done for some and it allows to lock the variables if needed later on.

I didn't really see a downside of using Getters and Setters here.

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -15,180 +15,189 @@
 
-	if addressIP.EndpointID() != 0 {
-		return nil, ErrNodeIPEndpointIDSet
+	if len(ipv6Address) != 16 {
```
should be `len(raw) != 16`, also be aware an `net.IPv4` address has also len 16

aanm replied:
addressed?

aanm commented on tests/manual-multi-node.bash:
```diff
@@ -67,8 +67,8 @@ function test_nodes {
 
-	node_run cilium-master "sudo cp /etc/init/cilium.conf tmp; sudo sed -i '/exec/d' tmp; echo \"exec cilium-agent --debug -n f00d::c0a8:210b:0:0 --ipv4-range 10.1.0.1 $OPTS\" | sudo tee -a tmp; sudo cp tmp /etc/init/cilium.conf; sudo service cilium restart"
-	node_run cilium-node-2 "sudo cp /etc/init/cilium.conf tmp; sudo sed -i '/exec/d' tmp; echo \"exec cilium-agent --debug -n f00d::c0a8:210c:0:0 --ipv4-range 10.2.0.1 $OPTS -c 192.168.33.11:8500\" | sudo tee -a tmp; sudo cp tmp /etc/init/cilium.conf; sudo service cilium restart"
+	node_run cilium-master "sudo cp /etc/init/cilium.conf tmp; sudo sed -i '/exec/d' tmp; echo \"exec cilium-agent --debug -n f00d::c0a8:210b:0:0 --ipv4-range 10.1.0.0/16 $OPTS\" | sudo tee -a tmp; sudo cp tmp /etc/init/cilium.conf; sudo service cilium restart"
```
Change the `daemon/main.go`
It is currently as this:
```
IPv4 allocation range for local endpoints, must be in correct format (10.N.0.1/16)
```

tgraf replied:
That change was in a later commit, I moved it to earlier now.

aanm commented on daemon/daemon.go:
```diff
@@ -607,7 +595,7 @@ func NewDaemon(c *Config) (*Daemon, error) {
 		policy:            policy.NewPolicyRepository(),
 		ignoredContainers: make(map[string]int),
-		buildEndpointChan: make(chan *endpoint.Request, common.EndpointsPerHost),
+		buildEndpointChan: make(chan *endpoint.Request, 512),
```
Why not `common.EndpointsPerHost`? With a queue of 512 There can be potentials deadlocks if you have >=512 endpoints and regenerate all endpoints at the same time

tgraf replied:
If this is the code then the surrounding code needs to be changed. We can have the same endpoint being rebuilt multiple times. I've bumped endpoints per host to 64K. We absolutely don't want a build backlog of 64K.

aanm replied:
>We can have the same endpoint being rebuilt multiple times.

We can but, for example, if you try to regenerate the same endpoint 100 times in 1 second, the queue will only contain a single entry of that endpoint, not 100 entries.
The entries in the queue are guarantee to be unique.

tgraf replied:
Discussed a bit on slack. I've put the original channel size in place again. We will look at this issue in a separate PR.

aanm commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -138,129 +166,56 @@ func (lxc LXCInfo) String() string {
-	)
-	if err != nil {
-		return nil, err
+func init() {
```
Init on top?

aanm commented on daemon/main.go:
```diff
@@ -244,7 +245,9 @@ func init() {
 		"enable-tracing", false, "Enable tracing while determining policy (debugging)")
 	flags.StringVar(&v4Prefix,
-		"ipv4-range", "", "IPv4 allocation range for local endpoints, must be in correct format (10.N.0.1/16)")
+		"ipv4-range", "", "Per-node IPv4 endpoint prefix (default: auto)")
```
If the default is auto why not set it up in the default value?
```
"ipv4-range", "auto", "Per-node IPv4 endpoint prefix")
```
Same with `ipv6-range`

tgraf replied:
Interesting idea. I like it.

aanm commented on daemon/main.go:
```diff
@@ -244,7 +245,9 @@ func init() {
-		"ipv4-range", "", "IPv4 allocation range for local endpoints, must be in correct format (10.N.0.1/16)")
+		"ipv4-range", "", "Per-node IPv4 endpoint prefix (default: auto)")
+	flags.StringVar(&v6Prefix,
+		"ipv6-range", "", "Per-node IPv6 endpoint prefix, must be /96 (default; auto)")
```
`s/(default; auto)/(default: auto)`

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -125,6 +125,14 @@ func GetIPv6ClusterRange() *net.IPNet {
 func GetIPv6AllocRange() *net.IPNet {
+	return &net.IPNet{
+		IP:   ipv6AllocRange.IP,
+		Mask: net.CIDRMask(IPv6NodeAllocPrefixLen, 128),
```
We have a mask for this, no? `StateIPv6Mask`

aanm commented on daemon/main.go:
```diff
@@ -280,8 +281,10 @@ func init() {
-	flags.StringVarP(&v6Address,
-		"node-address", "n", "", "IPv6 address of node, must be in correct format (XXXX:XXXX:XXXX:XXXX:XXXX:XXXX::/96)")
+	flags.StringVar(&v6Address,
+		"node-ipv6", "", "IPv6 address of node (default: auto)")
```
The same `auto` comments I've made in the previous commit.

Any constrains with this `node-ipv6`? Can be anything I want? Won't cilium fail if I assign a link-local `fe80` on purpose?

tgraf replied:
Good point regarding link local. I'll check for this.

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -28,12 +28,31 @@ import (
 )
 
+func makeIPv6RouterIP(ip net.IP) net.IP {
+	// Derive prefix::1 as the IPv6 pseudo router address
```
Did you forget to copy the IP? Before changing it?

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -28,12 +28,31 @@ import (
+}
+
+func makeIPv6HostIP(ip net.IP) net.IP {
+	// Derive prefix::2 as the IPv6 host address
```
Did you forget to copy the IP? Before changing it?

tgraf committed:
cli: Command to print contents of local tunnel endpoint map

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
vagrant: Use --ipv6-range to set per node prefix

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Populate local node in tunnel endpoint map

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Add --ipv6-range to set per node IPv6 allocation prefix

- Enforce strict prefix length of /96 per node and automatically
  derive the /112 as the allocation prefix from it.

- Report both as informational messages during bootstrapping.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
lxcmap: Support migration of map format

1. The endpoints map contains persistent information. Detect format
   inconsistentcy based on the map configuration in the code and the map
   information retrieved after opening the map.

2. If the map does not contain any entries, delete the map to force
   recreation in the new format.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Simplified encapsulation mode

- Use complete IP address when performing lookup in endpoint map. This
  allows to use arbitrary IP addresses for local endpoints without the
  constraints of a particular prefix.

- Use a new ENDPOINT_F_HOST flag to mark endpoints which represent the
  local host. This avoids multiple IPs to represent the local host.

- Instead of statically mapping from container IP to the remote tunnel
  endpoint, use a BPF map lookup for the mapping. This lessens the
  addressing constraints further and allows tunnel endpoints to have
  arbitrary addresses

  The key for the tunnel map lookup is the destination IP masked with
  the configured per node prefix length. This avoids requiring to
  populate the map with IPs of each individual endpoint. Later, we could
  use an LPM map for this to allow for exceptions to allow for mobility.

- The tunnel endpoint map automatically gets populated with all prefixes
  and node IPs received via node notifications (k8s/kvstore).

Fixes: #989
Fixes: #972

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Provide options to overwrite IPv4/IPv6 node address

Rename the existing -n/--node-addres into: --node--ipv4 and --node-ipv6
to overwrite the respective node addresses.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add NonPersistent flag to force removal of map on startup

This can be used to automatically upgrade the map format for maps which
never carry any persistent storage across agent restarts.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Print IPv4/IPv6 node IPs when starting up

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
daemon: Use real node address of host

So far, we have generated a special node IP based on the node prefix,
e.g. if a node was configured to use the prefix 10.6.0.0/16 we
automatically generated and used 10.6.0.1 to represent the node itself.
The user was then required to ensure that this node IP can be routed on
the network of the host. The main reason for this model was allow for
simple mapping of an IP to it's node IP.

With the endpoint table using complete IP addresses and a flag to
indicate node IPs. We can move to using the real, existing IP of the
node for this purpose:

- Derive global scope address of the host. If -d is provided, the first
  global scope address of that device is preferred, otherwise the first
  global scope adress of the system itself is chosen

- This derived node address is then being used throughout the agent,
  i.e. containers will have a default route pointing to this node
  address, the node address will be announced as the node address to
  other nodes for encapsulation, etc.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Split EndpointsCIDR into IPv4 and IPv6 field

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Derive real IPv6 node adress and split router address

- Derive the real IPv6 address of the node and use it if possible.
  Fall back to a generated address in the node prefix if not
  available (prefix::nodeid:0:1). Reserve this address in the
  allocation pool if needed.

- Continue generating a router IPv6 address for internal use but
  allocate one randomly from the pool to ensure it is unique.

- Fix BPF side to no longer use host IP when determing whether an
  address is inside the cluster prefix, use the router ip for this
  which always resides in the node prefix.

- Provide the ipv6 node prefix to the bpf/init.sh script

Fixes: #313

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Delay opening of maps to after mount

Schedule opening and creation of maps to after mounting has
occured.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests/lb: Do not rely on cilium_host having a global scope addres

Assign a temporary IPv6 address for the duration of the test and assume
that a global scope IPv4 address always exists.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests/proxy: Show return http codes

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
agent: Masquerade IPv4 packets which leave the host

All IPv4 packets which originated from the per node allocation prefix
and are leaving the host out a non tunnel interface will be
automatically masqueraded.

Masquerading can be disabled with --masquerade false or when enabling
direct routing mode (-d DEVICE).

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
tests/k8s: Adjust k8s multi node to new multi node mode

- Mount BPF filesystem when bootstrapping vagrant node
- Remove need for node specific DaemonSet files
- Remove NodeSelector code
- Remove invalid node-ip setting
- Add scheduler toleration to no longer require to initialize CNI manually
- Remove hardcoding of node address prefix
- Add controllerManagerExtraArgs to kubeadm config to enable per node
  CIDR allocation by Kubernetes
- Rename vagrant nodes to k8s1, k8s2
- Generalize Vagrantfile to allow going to more nodes easily

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on pkg/node/manager.go:
```diff
@@ -32,29 +37,64 @@ func GetNode(ni Identity) *Node {
+	}
+
+	if err := tunnel.DeleteTunnelEndpoint(ip.IP); err != nil {
+		log.Debugf("bpf: Unable to delete %s in tunnel endpoint map: %s", ip, err)
```
Small nit, why not throwing this error to the caller?

tgraf replied:
How do you want to handle the error?

aanm replied:
I was thinking on what we discussed on keeping a cache, this is fine then

tgraf replied:
OK, understood. We can add it when we need it.

aanm commented on pkg/node/manager.go:
```diff
@@ -32,29 +37,64 @@ func GetNode(ni Identity) *Node {
+		}
+	}
+
 	delete(nodes, ni)
```
Didn't we discuss that we wouldn't delete the nodes from the golang map in case the bpf throw an error?

tgraf replied:
Agreed. I think this should also use the state mechanism once we have it which then attempts to delete multiple times. I will change this to keep the node for now.

aanm commented on daemon/daemon.go:
```diff
@@ -520,50 +518,69 @@ func (d *Daemon) init() error {
+	if allocRange.Contains(nodeIP) {
+		err := ipamConf.IPv6Allocator.Allocate(nodeIP)
+		if err != nil {
+			log.Debugf("Unable to reserve IPv6 router address '%s': %s",
```
Should we ignore this? It's the router address where a container can potentially colide with this one, won't it will be able to receive all traffic from the remaining endpoints?

aanm commented on daemon/daemon.go:
```diff
@@ -625,16 +642,24 @@ func NewDaemon(c *Config) (*Daemon, error) {
 		}
 	}
 
+	if v4Prefix != "auto" {
```
small nit, replace `"auto"` with a const?

aanm commented on daemon/main.go:
```diff
@@ -244,7 +244,7 @@ func init() {
 		"enable-tracing", false, "Enable tracing while determining policy (debugging)")
 	flags.StringVar(&v4Prefix,
-		"ipv4-range", "", "IPv4 allocation range for local endpoints, must be in correct format (10.N.0.1/16)")
+		"ipv4-range", "auto", "Per-node IPv4 endpoint prefix")
```
Please write an example of the expected `ipv4-range`.

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -15,180 +15,189 @@
+// scope will be regarded as the system's node address.
+func InitDefaultPrefix(device string) {
+	ip, err := firstGlobalV4Addr(device)
+	if err == nil {
```
small nit:
```
if err != nil{
    log.Debugf("error...")
    return
}
// rest of the code
```

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -15,180 +15,189 @@
+			log.Panicf("BUG: Invalid default prefix '%s': %s", v6range, err)
+		}
+
+		ip6[14] = 0xff
```
Don't we have a function that sets the IPv6 address with `ffff` at the end?

aanm commented on pkg/nodeaddress/node_address.go:
```diff
@@ -201,3 +204,26 @@ func GetIPv4NodeRoute() net.IPNet {
+
+// GetNode returns the identity and node spec for the local node
+func GetNode() (node.Identity, *node.Node) {
+	range4, range6 := ipv4AllocRange, ipv6AllocRange
```
I think what you want is 
```
range4, range6 := *ipv4AllocRange, *ipv6AllocRange
```
and on line 224
```
		IPv4AllocCIDR: &range4,
		IPv6AllocCIDR: &range6,
```

otherwise if you change `ipv4AllocRange` in the future, the value is also changed in `range4`

aanm commented on daemon/daemon.go:
```diff
@@ -127,14 +127,6 @@ func (d *Daemon) RemoveProxyRedirect(e *endpoint.Endpoint, l4 *policy.L4Filter)
 }
 
-func (d *Daemon) WriteEndpoint(e *endpoint.Endpoint) error {
-	if err := d.conf.LXCMap.WriteEndpoint(e); err != nil {
```
We don't need the lxcmap anymore?

tgraf replied:
lxcmap instance is now maintained in the package itself. There is always only a single instance.

aanm commented on pkg/bpf/map.go:
```diff
@@ -216,6 +217,57 @@ func (m *Map) setPathIfUnset() error {
 	return nil
 }
 
+func (m *Map) migrate(fd int) (bool, error) {
```
I'm just writing this as a precaution: between line 220 and 306 are any garbage of maps file descriptors left open?

ianvernon replied:
Nit: function name is a bit misleading; I would expect 'migrate' to migrate data from an old map to a new map, but this function checks for differences between the existing map information in the BPF FS and the information Cilium contains about said map without migrating any data between the old map and the new one. Maybe rename to 'resolveMapMismatch' or something more evocative of what the function does.

tgraf replied:
In the future, migrate can actually perform migration. We are not maintaining backwards compatibility yet though. Still confusing?

aanm commented on pkg/bpf/map.go:
```diff
@@ -234,11 +286,24 @@ func (m *Map) OpenOrCreate() (bool, error) {
 		os.Remove(m.path)
 	}
 
+reopen:
```
my previous comment is regarding this `goto`, the `fd` open on the next line doesn't seem to be properly closed (I think)

aanm commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -18,31 +18,39 @@ import (
+
+func init() {
+	if _, err := mapInstance.OpenOrCreate(); err != nil {
+		log.Errorf("Error while opening endpoint map: %s", err)
```
no need for `log.Fatal` here?

aanm commented on pkg/maps/lxcmap/lxcmap.go:
```diff
@@ -138,129 +172,50 @@ func (lxc LXCInfo) String() string {
-			From: common.Swab16(pM.From),
-			To:   common.Swab16(pM.To),
+	for _, k := range f.GetBPFKeys() {
+		if err := mapInstance.Update(k, *info); err != nil {
```
`FIXME: revert previous updates?`

aanm commented on daemon/daemon.go:
```diff
@@ -644,6 +646,17 @@ func NewDaemon(c *Config) (*Daemon, error) {
 		nodeaddress.SetIPv4AllocRange(net)
 	}
 
+	if v6Prefix != "auto" {
```
replace `"auto"` with const?

aanm commented on daemon/main.go:
```diff
@@ -245,6 +246,8 @@ func init() {
 	flags.StringVar(&v4Prefix,
 		"ipv4-range", "auto", "Per-node IPv4 endpoint prefix")
+	flags.StringVar(&v6Prefix,
+		"ipv6-range", "auto", "Per-node IPv6 endpoint prefix, must be /96")
```
What was the outcome of running it with a link-local ipv6 address does cilium handle it well?

also:
`"Per-node IPv6 endpoint prefix, must be /%d", IPv6NodePrefixLen`?

aanm commented on daemon/main.go:
```diff
@@ -280,8 +281,10 @@ func init() {
-	flags.StringVarP(&v6Address,
-		"node-address", "n", "", "IPv6 address of node, must be in correct format (XXXX:XXXX:XXXX:XXXX:XXXX:XXXX::/96)")
+	flags.StringVar(&v6Address,
+		"ipv6-node", "auto", "IPv6 address of node")
```
I would like to see an example of what the user should write here

aanm commented on daemon/main.go:
```diff
@@ -280,8 +281,10 @@ func init() {
+	flags.StringVar(&v6Address,
+		"ipv6-node", "auto", "IPv6 address of node")
+	flags.StringVar(&v4Address,
+		"ipv4-node", "auto", "IPv4 address of node")
```
I would like to see an example of what the user should write here

aanm commented on daemon/daemon.go:
```diff
@@ -296,6 +296,70 @@ func (d *Daemon) setHostAddresses() error {
+		"-X", "CILIUM_POST"}, true)
+}
+
+func (d *Daemon) installMasqRule() error {
```
this is a workaround right?

tgraf replied:
No. Do you have a better idea?

aanm commented on tests/k8s/multi-node/cluster/cluster-manager.bash:
```diff
@@ -319,76 +294,30 @@ function deploy_cilium(){
         kubectl create -f "${cilium_dir}"
 
-        wait_for_daemon_set_ready kube-system cilium-server-1 1
+        wait_for_daemon_set_ready kube-system cilium-lb 2
```
2 cilium-lb? The cilium-lb is a daemon-set to test k8s ingress with a specific node selector label. There should be a single ingress controller running in the k8s cluster otherwise we will end up with conflicting IPs when trying to update the frontend IP in the kubernetes ingress resource.

borkmann reviewed (approved):
LGTM, would be great if we could get rid of masquerade target at some point later and also do this in BPF.

ianvernon reviewed (approved):
Reviewed the Go code - looks great. Have one small nitpick which does not need a second review.
